File: main/algorithms/arrays/sorting/insertion/go/insertion.go
/*
    type: algorithm
    theme: arrays
    sub-theme: sorting
    name: insertion sorting
    author of code: Evgeny @missingdays Bovykin

*/

package main

import "fmt"

//Sort given array using insertion sort algorithm
func insertionSort(a []int) {

    //For every element in array
    for i := 1; i < len(a); i++ {

        //Get its value
        value := a[i]

        //Decrease iterator by one
        j := i - 1

        //Insert element to a place where it belongs
        for j >= 0 && a[j] > value {

            //shifting all elements after it to the right by one
            a[j+1] = a[j]
            j = j - 1
        }

        //Write number to place where it should be
        a[j+1] = value
    }
}

func main() {

    //Array to be sorted
    list := []int{31, 41, 59, 26, 53, 58, 97, 93, 23, 84}
    fmt.Println("unsorted:", list)

    //Sort it!
    insertionSort(list)
    fmt.Println("sorted!  ", list)
}

File: nerdy/algorithms/sort/insertion/insertion.py
'''
    type: algorithm
    theme: arrays
    sub-theme: sorting
    name: insertion sorting
    author of code: Evgeny @missingdays Bovykin

'''

def insertionSort(ar):

    for i in range(1, len(ar)):

        #Get new element
        current_value = ar[i]
        #Get position of this element
        position = i

        #While new element is bigger than compared value
        while(position > 0 and ar[position-1] > current_value):

            #Dive this element down
            ar[position] = ar[position-1]
            position = position - 1

        #Write element to its new place
        ar[position] = current_value

#Array we will sort
ar = [0, 1, 10, 15, 7, 20, 4, 6, 90, 5]

#Sort it
insertionSort(ar)

#Print it
for i in range(len(ar)):
    print(ar[i])
